include/apq/impl/async_connect.h

```diff
@@ -0,0 +1,172 @@
+#pragma once
+
+#include <apq/connection.h>
+#include <apq/error.h>
+#include <apq/detail/bind.h>
+#include <libpq-fe.h>
+
+namespace libapq {
+namespace impl {
+namespace pq {
+
+template <typename T, typename Handler, typename = Require<Connectiable<T>>>
+inline void pq_write_poll(T& conn, Handler&& h) {
+    get_socket(conn).async_write_some(
+            asio::null_buffers(), std::forward<Handler>(h));
+}
+
+template <typename T, typename Handler, typename = Require<Connectiable<T>>>
+inline void pq_read_poll(T& conn, Handler&& h) {
+    get_socket(conn).async_read_some(
+            asio::null_buffers(), std::forward<Handler>(h));
+}
+
+template <typename T, typename = Require<Connectiable<T>>>
+inline decltype(auto) pq_connect_poll(T& conn) {
+    return PQconnectPoll(get_native_handle(conn));
+}
+
+template <typename T, typename = Require<Connectiable<T>>>
+inline error_code pq_start_connection(T& conn, const std::string& conninfo) {
+    pg_conn_handle handle(PQconnectStart(conninfo.c_str()), PQfinish);
+    if (!handle) {
+        return make_error_code(error::pq_connection_start_failed);
+    }
+    get_handle(conn) = std::move(handle);
+    return {};
+}
+
+template <typename T, typename = Require<Connectiable<T>>>
+inline error_code pq_assign_socket(T& conn) {
+    int fd = PQsocket(get_native_handle(conn));
+    if (fd == -1) {
+        return make_error_code(error::pq_socket_failed);
+    }
+
+    int new_fd = dup(fd);
+    if (new_fd == -1) {
+        using namespace boost::system;
+        return make_error_code(error_code{errno, get_posix_category()},
+            "dup(fd)");
```

Comment:
```"dup(fd)``` is not very informative message. Use ```std::strerror(errno)```.

Reply:
Well, std::strerror(errno) is made inside posix_category::message, so "dup(fd)" gives an information about where the error occurred. The message can be changed to contain more words if it really needed.

Reply:
ok